src/Firestore/ValueMapper.php

```diff
@@ -0,0 +1,503 @@
+    {
+        $output = [];
+
+        foreach ($fields as $key => $val) {
+            $type = array_keys($val)[0];
+            $value = current($val);
+
+            $output[$key] = $this->decodeValue($type, $value);
+        }
+
+        return $output;
+    }
+
+    /**
+     * Convert a PHP array containing google-cloud-php and simple types to an
+     * array ready to be sent to Firestore.
+     *
+     * @param array $fields A list of fields to encode.
+     * @return array
+     */
+    public function encodeValues(array $fields)
+    {
+        $output = [];
+
+        foreach ($fields as $key => $val) {
+            $output[$key] = $this->encodeValue($val);
+        }
+
+        return $output;
+    }
+
+    /**
+     * Escape a field path and return it as a string.
+     *
+     * @param string|FieldPath $fieldPath
+     * @return string
+     * @throws \InvalidArgumentException If the path is a string, and is invalid.
+     */
+    public function escapeFieldPath($fieldPath)
+    {
+        if ($fieldPath instanceof FieldPath) {
+            $parts = $fieldPath->path();
+
+            $out = [];
+            foreach ($parts as $part) {
+                $out[] = $this->escapePathPart($part);
+            }
+
+            $fieldPath = implode('.', $out);
+        }
+
+        $this->validateFieldPath($fieldPath);
+        return $fieldPath;
+    }
+
+    /**
+     * Create a list of fields paths from field data.
+     *
+     * The return value of this method does not include the field values. It
+     * merely provides a list of field paths which were included in the input.
+     *
+     * @param array $fields A list of fields to map as paths.
+     * @param string $parentPath The parent path (used internally).
+     * @return array
+     */
+    public function encodeFieldPaths(array $fields, $parentPath = '')
+    {
+        $output = [];
+
+        foreach ($fields as $key => $val) {
+            $key = $this->escapePathPart($key);
+
+            if (is_array($val) && $this->isAssoc($val)) {
+                $nestedParentPath = $parentPath
+                    ? $parentPath . '.' . $key
+                    : $key;
+
+                $output = array_merge($output, $this->encodeFieldPaths($val, $nestedParentPath));
+            } else {
+                $output[] = $parentPath
+                    ? $parentPath . '.' . $key
+                    : $key;
+            }
+        }
+
+        return $output;
+    }
+
+    /**
+     * Accepts a list of FieldPath objects and a list of values, and constructs
+     * a nested array of fields and values.
+     *
+     * @param FieldPath[] $paths The field paths.
+     * @param array $values The field values.
+     * @return array
+     * @todo less recursion
+     */
+    public function buildDocumentFromPathsAndValues(array $paths, array $values)
+    {
+        $this->validateBatch($paths, FieldPath::class);
+
+        $output = [];
+
+        foreach ($paths as $pathIndex => $path) {
```

Comment:
What does this function do with overlapping field paths, such as 'a.b':'foo', 'a':'bar'?

Reply:
If the paths are mutually exclusive (as they are in your example, as `a` cannot be both a string and map value), the latter value will overwrite the former.

Reply:
```php
$document->update([
    ['path' => new FieldPath(['a', 'b']), 'value' => 'foo'],
    ['path' => new FieldPath(['a']), 'value' => 'bar'],
]);
```

Results in:
```
Array
(
    [a] => bar
)
```

Reply:
We should return an error here (but we can do this as a follow-up).